c10/core/TensorImpl.h

```diff
@@ -2932,6 +2948,17 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   DispatchKeySet key_set_;
 
  private:
+  intrusive_ptr<impl::CopyOnWriteSimulator> copy_on_write_simulator_;
```

Comment:
extra refcounted field, bloop.

Reply:
Is there any way around that for the lifetime of this warning?

Reply:
I think we can mitigate the impact of this by delaying population of this in the tensor until an actual copy-on-write is performed. Basically, we move this into the storage, and the storage will also track whether or not a copy-on-write has been performed. If we are in a situation where the storage reports that there's been a copy-on-write, but the tensor doesn't have a simulator, then it knows that it can inherit the "default" simulator since it is the simulator that all views would have had if we did this eagerly.

We can apply the same logic to defer any generation tracking as well.

Reply:
If I am understanding this specific proposal correctly, this could also allow us to justify paying for a mutex to do this check with higher fidelity (synchronizing reads and writes to the generation numbers), since we ought to only need to use the mutex in situations where a copy-on-write would have actually been performed.

We can also throw all this checking away once we've warned once, since we'll never warn again.

So the state diagram I imagine is this:
1) tensor is created. views of this tensor may also be created. There's a single storage, and it says that there's no copy-on-writes going on. Each tensor has a null copy-on-write simulator.
2) reshape would have created a CoW tensor. The new "copy" has the aliased storage, but it also has its own CoW simulator inside the tensor. We create a default CoW simulator inside the storage: this default value is only used by all of the views that pre-existed the copy-on-write.
2b) all read/write operations at this point to any of the view families that share this storage are now going to do all the fun generation tracking.
3) if at any point we trigger the warning, then we can discard all the state for this storage. If we want to go further, we can set a global telling us to never create any copy-on-write state again, assuming we are doing a TORCH_WARN_ONCE.

Note that we may want to distinguish reads after writes to a different copy and writes after writes to a different copy (i.e. two different warnings).

Reply:
OK, unfortunately we're a little late for that. I have implemented this but it does still require a mutex acquisition even when we aren't using this. That can probably be downgraded to an atomic read eventually.

Testing this new configuration in CI, here we go!

Reply:
OK, seems like we're ok with the approach here. We can drop this to not be refcounted as described in the conversation about templating on intrusiveness.